### PR TITLE
Utilize graphql's provided validations

### PR DESCRIPTION
**Pull request by stefanpenner**
This demonstrates it is not only possible, but totally viable. Next steps are:

* [x] demonstrate viability
* [x] is to clean this up
* [x] make configuration work properly
* [x] test thoroughly
* [x] see how it works with importable fragments

image

### CODE REVIEW

File: projects/eslint-plugin-graphql/eslint/graphql-rule-adapter.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    create(context) {
+      const wrappedRules = Object.create(null);
+      const typeInfo = new TypeInfo(context.parserServices.getSchema());
+      const validationContext = new ValidationContext(
+        context.parserServices.getSchema(),
+        context.parserServices.getDocument(),
+        typeInfo,
+        function onError(error) {
+          const errorNode = error.nodes[0];
+          const node = context.parserServices.correspondingNode(errorNode) || {
```
Comment (stefanpenner) [inline projects/eslint-plugin-graphql/eslint/graphql-rule-adapter.js:41]: TODO: Lock this down

File: projects/eslint-plugin-graphql/eslint/parser.js [hunk 2]
```diff
@@ -1,65 +1,114 @@
...
-const gqlToEs = require('./gql-node-to-es-node');
+const { parse, buildSchema } = require('graphql');
+const Traverser = require('eslint/lib/shared/traverser');
+const path = require('path');
+const fs = require('fs');
 
-module.exports.parseForESLint = function(code, options) {
-  let ast;
+const toEslintAST = require('./to-eslint-ast');
+// TODO: verify this is the required set. Note: if something isn't here, it
```
Comment (stefanpenner) [inline projects/eslint-plugin-graphql/eslint/parser.js:9]: TODO: lock this down

File: projects/eslint-plugin-graphql/eslint/parser.js [hunk 3]
```diff
@@ -1,65 +1,114 @@
...
-        });
-        return ast;
+      getSchema() {
+        return schema;
+      },
+      getDocument() {
+        return this.correspondingNode(ast);
+      },
+      parse(source) {
+        // TODO:parse with options?
```
Comment (stefanpenner) [inline projects/eslint-plugin-graphql/eslint/parser.js:75]: TODO: lock this down

File: projects/eslint-plugin-graphql/eslint/rules/recommended.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+module.exports = {
```
Comment (stefanpenner) [inline projects/eslint-plugin-graphql/eslint/rules/recommended.js:1]: Don't duplicate this file

File: projects/eslint-plugin-css/tests/parser-test.js [hunk 5]
```diff
@@ -9,21 +9,13 @@ describe('parser', function() {
     expect(parser.parseForESLint).to.be.a('function');
   });
 
-  it('provides a good error if extension could not parse', function() {
+  it.skip('provides a good error if extension could not parse', function() {
```
Comment (hjdivad) [inline projects/eslint-plugin-css/tests/parser-test.js:12]: unskip?

File: projects/eslint-plugin-example/tests/parser-test.js [hunk 6]
```diff
@@ -9,7 +9,7 @@ describe('parser', function() {
     expect(parser.parseForESLint).to.be.a('function');
   });
 
-  it('provides a good error if extension could not parse', function() {
+  it.skip('provides a good error if extension could not parse', function() {
```
Comment (hjdivad) [inline projects/eslint-plugin-example/tests/parser-test.js:12]: unskip?

File: projects/eslint-plugin-graphql/README.md [hunk 7]
```diff
@@ -19,21 +20,79 @@ module.exports = {
...
+```
+# eslint-disable <rule-name>, <rule-2-name>, ... <rule-n-name>
+# eslint-enable <rule-name>, <rule-2-name>, ... <rule-n-name>
+# eslint-disable-line
+# eslint-disable-line <rule-name>, <rule-2-name>, ... <rule-n-name>
+# eslint-disable-next-line
+# eslint-disable-next-line <rule-name>, <rule-2-name>, ... <rule-n-name>
+```
+
+For our built-in-rules, they will have the name structure:
```
Comment (hjdivad) [inline projects/eslint-plugin-graphql/README.md:46]:
```suggestion
For our built-in rules, they will have the name structure:
```

File: projects/eslint-plugin-graphql/README.md [hunk 8]
```diff
@@ -19,21 +20,79 @@ module.exports = {
...
+
 ## vim-coc users
 
 1. open `:CocConfig`
 2. add `'graphql'` to `'eslint.filetypes'`
 
 
+## List Of Validations:
+
+This project projects some of it's own rules:
```
Comment (hjdivad) [inline projects/eslint-plugin-graphql/README.md:66]:
```suggestion
This project includes some of its own rules:
```

File: projects/eslint-plugin-graphql/README.md [hunk 9]
```diff
@@ -19,21 +20,79 @@ module.exports = {
...
+
+* single-top-level-query
+
+This project supports and executes all graphql's own validations as of version 15.3.0, they are:
```
Comment (hjdivad) [inline projects/eslint-plugin-graphql/README.md:70]:
```suggestion
This project supports and executes all of graphql's own validations as of version 15.3.0; they are:
```

File: projects/eslint-plugin-graphql/eslint/parser.js [hunk 10]
```diff
@@ -1,65 +1,114 @@
...
+
+  if (options.schema) {
+    // if we are given an absolute path, automatically just read it
+    if (path.isAbsolute(options.schema)) {
+      schemaString = fs.readFileSync(options.schema, 'UTF8')
+    } else {
+      // otherwise resolve, that enables us to resolve relative the path of the file
+      schemaString = fs.readFileSync(resolve(options.schema, {
+        basedir: path.dirname(options.filePath)
+      }), 'UTF8');
```
Comment (hjdivad) [inline projects/eslint-plugin-graphql/eslint/parser.js:45]: require.resolve

File: projects/eslint-plugin-graphql/tests/parser-test.js [hunk 11]
```diff
@@ -50,6 +51,55 @@ describe('parser', function() {
...
-    expect(result).to.have.keys(['ast', 'scopeManager', 'visitorKeys']);
+    expect(result).to.have.keys(['ast', 'scopeManager', 'visitorKeys', 'services']);
+  });
+
+  describe('parser services', function() {
+    it('has a functioning parse method', function() {
+      const source = `
+query {
+  id
+}
```
Comment (hjdivad) [inline projects/eslint-plugin-graphql/tests/parser-test.js:62]: https://github.com/zspecza/common-tags#stripindenttransformertypeinitial

**Review by hjdivad — CHANGES_REQUESTED** [review]

**Review by rwjblue — COMMENTED** [review]
* Can you make sure that we have this library under lint? During review, it seems that there might have been things that _should_ have failed linting but didn't. Good to test...
* There are a few different files where it seems that the indentation is a bit odd/off, might be worth fixing up...

File: projects/eslint-plugin-graphql/eslint/parser.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (stefanpenner) [reply projects/eslint-plugin-graphql/eslint/parser.js:45]: david convinced us not too

File: projects/eslint-plugin-graphql/tests/parser-test.js (hunk 11 above)
Comment (stefanpenner) [reply projects/eslint-plugin-graphql/tests/parser-test.js:62]: cool library, next time i'll use that. Today switching to that, results in a decent number of test changes due to row/column changes. I'm not sure it's worth it right now, i have though tried to make them all consistent

**Comment by stefanpenner** [comment]
@rwjblue 
> There are a few different files where it seems that the indentation is a bit odd/off, might be worth fixing up...

Should be more consistent now, I just opted for always ending the string with [`,] and not leaving dangling whitespace in the spring

File: projects/eslint-plugin-graphql/eslint/parser.js (hunk 10 above)
Comment (stefanpenner) [reply projects/eslint-plugin-graphql/eslint/parser.js:45]: what do we resolve relative to?

Comment (hjdivad) [reply projects/eslint-plugin-graphql/eslint/parser.js:45]: throwing a nice error that tells the user to put `${__dirname__}/foo/bar` would be fine too.  The main thing is I expect users to configure `eslintrc` with relative paths and if they need to do something else, the dx for this case should be a "pit of success" style.

**Comment by hjdivad** [comment]
things to consider
- [ ] what is the dx if you forget to specify `parser` in `eslintrc`?  can we detect & throw a useful error here

**Comment by hjdivad** [comment]
> what is the dx if you forget to specify parser in eslintrc? can we detect & throw a useful error here

Add troubleshooting to README; we can't do anything in code for this particular configuration problem

**Comment by hjdivad** [comment]
> error if an old version of eslint is used that doesn't support these capabilities.

can't give a good error here due to npm issues; updated readme

**Review by hjdivad — APPROVED** [review]

File: projects/eslint-plugin-graphql/eslint/parser.js [hunk 12]
```diff
@@ -1,65 +1,114 @@
...
+const visitorKeys = {
+  "Document": [ "definitions" ],
+  "FragmentDefinition": ["typeCondition","directives","selectionSet"],
+  "NamedType": [],
+  "SelectionSet": ["selections"],
+  "OperationDefinition": ["variableDefinitions","directives","selectionSet"],
+  "Field": ["alias","arguments","directives","selectionSet"],
+  "FragmentSpread": ["directives"],
+  "FieldDefinition": ["name", "arguments", "directives"]
+};
```
[the remark this answers is not in the document]
Comment (stefanpenner) [reply projects/eslint-plugin-graphql/eslint/parser.js]: Likely: `require('graphql/language/visitor').QueryDocumentKeys`

**Pull request merged**